// task/index.js
"use strict";

var fiber = require("fibers"),
    cond = require("cond"),
    Q = require("q");

function spawn(cb) {
  /*jshint validthis: true*/
  let cbArgs = [].slice.call(arguments, 1),
      deferred = Q.defer();
  function execCallback() {
    cond.handlerCase(function() {
      deferred.resolve(cb.apply(this, arguments));
    }, deferred.reject.bind(deferred));
  }
  let fib = fiber(execCallback.bind.apply(execCallback, [this].concat(cbArgs)));
  process.nextTick(fib.run.bind(fib));
  return deferred.promise;
}

function _yield(val) {
  let curr = fiber.current,
      e = {};
  if (typeof val.then === "function") {
    val.then(function(res) {
      curr.run(res);
    }, function(err) {
      e.err = err;
      curr.run(e);
    });
  } else {
    setTimeout(function() {
      curr.run(val);
    }, 0);
  }
  let res = fiber.yield();
  if (res === e) {
    return cond.error(res.err);
  } else {
    return res;
  }
}

function sleep(ms) {
  let curr = fiber.current;
  setTimeout(function() {
    curr.run();
  }, ms);
  fiber.yield();
}

function wrap(fn) {
  return function() {
    return _yield(fn.apply(this, arguments));
  };
}

function isInTask() {
  return typeof fiber.current !== "undefined";
}

module.exports = {
  spawn: spawn,
  yield: _yield,
  sleep: sleep,
  wrap: wrap,
  isInTask: isInTask
};
